053-configurable-event-store/proposal.md

```diff
@@ -0,0 +1,452 @@
+    `requested <= x <= n`). If the backend store doesn't provide an easy way to
+    read exactly `requested` elements, and the `EventStore` reads in a chunk of
+    data that results in more than `requested` events, there's no use in throwing
+    the extra events away.
+
+* `Delete(ctx, builds []db.Builds)` deletes the build events for a list of builds.
+  This is used by the build log collector.
+
+* `DeletePipeline(ctx, pipeline db.Pipeline)` deletes all of the builds for a given
+  pipeline. While this could be done by doing something like:
+
+  ```go
+  builds, _ := pipeline.Builds()
+  eventStore.Delete(builds)
+  ```
+
+  ...this doesn't work so well for the Postgres implementation. Currently, when
+  we delete a Pipeline from the database, `TRIGGER`s will `DROP` the
+  `pipeline_build_events_x` table for the appropriate pipeline. This results in
+  much more performant build event deletions, since dropping a table is faster
+  than deleting many thousands/millions of rows. I'd be happy to hear any clever
+  ideas anyone has on how to just have a single `Delete` method that acts on
+  builds while achieving the same table-dropping ability for Postgres.
+
+* `DeleteTeam(ctx, team db.Team)` deletes all builds for a given team. Same
+  reasoning as `DeletePipeline()`.
+
+* `UnmarshalKey(data []byte, key *Key)` works like `UnmarshalJSON` - give it
+  some bytes and tell it where you want to put the result. It is the inverse of
+  `Key.Marshal` for the corresponding `Key` type.
+
+
+## <a name="event-store-composition">`EventStore` Composition</a>
+
+In addition to enabling Concourse to interact with external stores besides
+Postgres, we can construct `EventStore` implementations that compose other
+`EventStores` to do some interesting things. One application is data
+transformations. I'll take the example of secret redaction:
+
+```go
+type SecretRedactingEventStore struct {
+    EventStore
+    // Some cred vars tracking state
+    ...
+}
+
+func (s *SecretRedactingEventStore) Initialize(ctx context.Context, build db.Build) error {
+    s.initializeCredVarsTrackingForBuild(build)
+    return s.EventStore.Initialize(ctx, build)
+}
+
+func (s *SecretRedactingEventStore) Put(ctx context.Context, build db.Build, events []atc.Event) error {
+    redactedEvents := make([]atc.Event, len(events)
+    for i, evt := range events {
+        redactedEvents[i] = s.redactSecrets(build, evt)
+    }
+    return s.EventStore.Put(ctx, build, redactedEvent)
+}
+```
+
+Another application is continuous syslog forwarding:
+
+```go
+type SyslogForwardingEventStore struct {
+    EventStore
+    // Some syslog forwarding state
+}
+
+func (s *SyslogForwardingEventStore) Put(ctx context.Context, build db.Build, events []atc.Event) error {
+    if err := s.syslogForwardEvents(build, events); err != nil {
+        return err
+    }
+    return s.EventStore.Put(ctx, build, event)
+}
+```
+
+How about migrating build events to cold-storage after they're completed:
```

Comment:
I wonder if the desire to design for cold storage doesn't mean that `EventStore.Get` should be some kind of optional operation. Usually the trade-off with cold storage is that it reduces storage costs while increasing access costs, and we wouldn't want to subject users to a surprisingly high bill just because they accessed some part of Concourse in a way which invoked the `EventStore.Get` on cold-storage.

Maybe a better design would be one where `EventStore.Get` can return a link to where the build events can be found? Then the user can more intentionally make the decision to access cold storage and invoke those costs.

Reply:
Hmm good point. I'm hesitant to modify the `EventStore` interface to explicitly make `Get` optional, since that seems pretty specific to this use case (and I think it generally makes sense for there to be `Get` functionality).

I guess this is also modifying the API (albeit in a different way), but we could introduce a special build event type that includes the link to cold-storage (e.g. `event.InColdStorage`), and a ColdStorage implementation would return that single build event when called. The event consumers (the UI and Fly) could know to interpret that event type specially, perhaps prompting the user to confirm they want to fetch from cold storage. This prompt could also be bypassed with a special `Key` (passed as the `Last-Event-ID` to the build events API endpoint) to indicate "yes, I know I'm fetching from cold storage, and I'm okay with that".